File: day10_builder/guru/cpp/main.cpp
// It makes sense to use the Builder pattern only when your products are quite
// complex and require extensive configuration.

#include <memory>

#include "builder.hpp"
#include "director.hpp"
#include "product.hpp"

// The client code creates a builder object, passes it to the director and then
// initiates the construction process. The end result is retrieved from the
// builder object.
void ClientCode(Director& director) {
  auto builder = std::make_shared<ConcreteBuilder>();
  director.SetBuilder(builder);

  director.BuildMinimalViableProduct();
  std::unique_ptr<Product> min_product = builder->GetProduct();
  min_product->ListParts();

  director.BuildFullFeaturedProduct();
  std::unique_ptr<Product> full_product = builder->GetProduct();
  full_product->ListParts();

  // Remember, the Builder pattern can be used without a Director class.
  std::cout << "Building customized product..." << '\n';
  builder->ProducePartA();
  builder->ProducePartC();
  std::cout << "Done." << '\n';
  std::unique_ptr<Product> custom_product = builder->GetProduct();
  custom_product->ListParts();
}


int main(int argc, char const *argv[]) {
  Director director;
  ClientCode(director);
  // Builder: Ready for a new build.
  // Director: Building basic product...
  // Done.
  // Builder: Ready for a new build.
  // Product parts: PartA
  //
  // Director: Building full featured product...
  // Done.
  // Builder: Ready for a new build.
  // Product parts: PartA, PartB, PartC
  //
  // Building customized product...
  // Done.
  // Builder: Ready for a new build.
  // Product parts: PartA, PartC

  return 0;
}

File: day10_builder/guru/py/main.py
from builder import *
from director import *
from product import *


def client_code(director: Director) -> None:
    """
    The client code creates a builder object, passes it to the director and then
    initiates the construction process. The end result is retrieved from the
    builder object.
    """

    builder = ConcreteBuilder1()
    director.builder = builder

    director.build_minimal_viable_product()
    builder.product.list_parts()
    print()

    director.build_full_featured_product()
    builder.product.list_parts()
    print()

    # Remember, the Builder pattern can be used without a Director class.
    print('Building customized product...')
    builder.produce_part_a()
    builder.produce_part_b()
    print('Done.')
    builder.product.list_parts()


if __name__ == '__main__':
    director = Director()
    client_code(director)
    # ConcreteBuilder1: Ready for a new build.
    # Director: Building basic product...
    # Done.
    # ConcreteBuilder1: Ready for a new build.
    # Product parts: PartA1
    #
    # Director: Building full featured product...
    # Done.
    # ConcreteBuilder1: Ready for a new build.
    # Product parts: PartA1, PartB1, PartC1
    #
    # Building customized product...
    # Done.
    # ConcreteBuilder1: Ready for a new build.
    # Product parts: PartA1, PartB1
